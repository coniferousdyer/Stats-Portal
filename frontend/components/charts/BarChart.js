// External ibrary components
import dynamic from "next/dynamic";
import Link from "next/link";
import { useState, useEffect } from "react";

// Material UI components
import Paper from "@mui/material/Paper";
import Button from "@mui/material/Button";

// CSS styles
import styles from "../../styles/components/charts/BarChart.module.css";

// Dynamic import that fixes the "ReferenceError: window is not defined" error
const Chart = dynamic(() => import("react-apexcharts"), { ssr: false });

// The data passed in, i.e. "data", must be of the form [[name_1, object_1], [name_2, object_2], ...]
const BarChart = ({
  title,
  horizontal,
  dataList,
  color,
  buttonText,
  buttonLink,
}) => {
  // The data series supplied to the chart
  const [series, setSeries] = useState([]);
  // The chart configuration options
  const [options, setOptions] = useState({
    chart: {
      type: "bar",
    },
    plotOptions: {
      bar: {
        borderRadius: 4,
        horizontal: horizontal,
      },
    },
    dataLabels: {
      enabled: false,
    },
    responsive: [
      {
        breakpoint: 600,
        options: {
          chart: {
            height: 750,
          },
          plotOptions: {
            bar: {
              horizontal: true,
            },
          },
        },
      },
    ],
  });

  // Modify the passed data into a format that ApexCharts can understand
  useEffect(() => {
    // Obtain the sorted set of all keys in the data
    let names = [];
    names = [
      ...new Set(...names, ...dataList.map((data) => Object.keys(data.series))),
    ];

    const dataSeries = dataList.map((data) => {
      return {
        name: data.name,
        data: names.map((name) => data.series[name] || 0),
      };
    });

    // Update the series
    setSeries(dataSeries);

    // Update the chart configuration with the new names and color, if defined
    setOptions({
      ...options,
      ...(color && { colors: color }),
      xaxis: {
        categories: names,
      },
    });
  }, [dataList]);

  return (
    <Paper elevation={5} style={{ height: "100%" }}>
      <h1 className={styles.title}>{title.toUpperCase()}</h1>
      <Chart options={options} series={series} type="bar" />
      {buttonText && (
        <div className={styles.button_container}>
          <Link href={buttonLink ? buttonLink : ""} passHref>
            <Button variant="contained" className={styles.button}>
              {buttonText}
            </Button>
          </Link>
        </div>
      )}
    </Paper>
  );
};

export default BarChart;
