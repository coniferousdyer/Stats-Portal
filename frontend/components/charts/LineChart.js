// External ibrary components
import dynamic from "next/dynamic";
import { useState, useEffect } from "react";

// Material UI components
import Paper from "@mui/material/Paper";

// CSS styles
import styles from "../../styles/components/charts/LineChart.module.css";

// Dynamic import that fixes the "ReferenceError: window is not defined" error
const Chart = dynamic(() => import("react-apexcharts"), { ssr: false });

// dataList is an array of arrays. Each array corresponds to a particular user.
// Each member of the dataList passed in must be of the form [{ x: value_x, y: value_y }, ...]
const LineChart = ({ title, dataList }) => {
  // The data series supplied to the chart
  const [series, setSeries] = useState([]);
  // The chart configuration options
  const [options, setOptions] = useState({
    chart: {
      type: "line",
      zoom: {
        enabled: true,
      },
    },
    dataLabels: {
      enabled: false,
    },
    stroke: {
      curve: "straight",
      width: 3,
    },
    grid: {
      row: {
        colors: ["#f3f3f3", "transparent"],
        opacity: 0.5,
      },
    },
    xaxis: {
      type: "datetime",
    },
    responsive: [
      {
        breakpoint: 900,
        options: {
          chart: {
            height: 350,
          },
          stroke: {
            width: 2,
          },
        },
      },
    ],
  });

  // Modify the passed data into a format that ApexCharts can understand
  useEffect(() => {
    const dataSeries = dataList.map((data) => {
      // data is of the form [{ x: value_x, y: value_y }, ...]. While this is a valid
      // format for the line chart, we need to first convert each object to the format
      // { x: (value_x in type datetime), y: value_y } for the time series to be valid.
      const ratingHistory = data.series.map((rating) => ({
        x: new Date(rating.date).getTime(),
        y: rating.rating,
      }));

      return {
        name: data.name,
        data: ratingHistory,
      };
    });

    // Update the series
    setSeries(dataSeries);
  }, [dataList]);

  return (
    <Paper elevation={5} style={{ height: "100%" }}>
      <h1 className={styles.title}>{title.toUpperCase()}</h1>
      {/* height is set to 100% here to limit the height of the chart */}
      <Chart options={options} series={series} type="line" height="100%" />
    </Paper>
  );
};

export default LineChart;
