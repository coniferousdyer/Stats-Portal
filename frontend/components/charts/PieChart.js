// External library components
import dynamic from "next/dynamic";
import { useState, useEffect } from "react";

// Material UI components
import Paper from "@mui/material/Paper";

// CSS styles
import styles from "../../styles/components/charts/PieChart.module.css";

// Dynamic import that fixes the "ReferenceError: window is not defined" error
const Chart = dynamic(() => import("react-apexcharts"), { ssr: false });

// The data passed in, i.e. "data", must be of the form { name_1: value_1, name_2: value_2, ... }
const PieChart = ({ title, donut, data }) => {
  // The data series supplied to the chart
  const [series, setSeries] = useState([]);
  // The chart configuration options
  const [options, setOptions] = useState({
    chart: {
      type: donut ? "donut" : "pie",
    },
    legend: {
      show: true,
      onItemHover: {
        highlightDataSeries: true,
      },
      onItemClick: {
        toggleDataSeries: true,
      },
    },
    stroke: {
      width: 0,
    },
    responsive: [
      {
        breakpoint: 480,
        options: {
          plotOptions: {
            pie: {
              customScale: 1.5,
              offsetY: 50,
            },
          },
          dataLabels: {
            enabled: false,
          },
          legend: {
            position: "bottom",
            offsetY: 50,
          },
        },
      },
    ],
  });

  // Modify the passed data into a format that ApexCharts can understand
  useEffect(() => {
    // data is of the form {name: value}. We need to sort this dictionary
    // by the count and then convert it into 2 arrays: names and values.
    const names = Object.keys(data).sort((a, b) => data[b] - data[a]);
    const values = Object.values(data).sort((a, b) => b - a);

    // Update the series
    setSeries(values);

    // Add the newly created labels to the configuration options
    setOptions({
      ...options,
      labels: names,
    });
  }, [data]);

  return (
    <Paper elevation={5} style={{ height: "100%" }}>
      <h1 className={styles.title}>{title.toUpperCase()}</h1>
      <Chart options={options} series={series} type={donut ? "donut" : "pie"} />
    </Paper>
  );
};

export default PieChart;
