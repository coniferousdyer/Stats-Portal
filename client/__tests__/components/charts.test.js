// For components imported using "next/dynamic".
import preloadAll from "jest-next-dynamic";

// External library components.
import Link from "next/link";
import Chart from "react-apexcharts";

// Material UI components.
import Button from "@mui/material/Button";

// External testing library components.
import TestRenderer from "react-test-renderer";

// Internal application components to be tested.
import BarChart from "../../components/charts/BarChart";
import LineChart from "../../components/charts/LineChart";
import PieChart from "../../components/charts/PieChart";

beforeAll(async () => {
  await preloadAll();
});

// BarChart tests.
describe("BarChart", () => {
  it("renders correctly with vertical bar chart", () => {
    const barChart = TestRenderer.create(
      <BarChart
        title="Test Title"
        horizontal={false}
        dataList={[
          {
            name: "Test Series 1",
            series: {
              A: 10,
              B: 20,
              C: 30,
            },
          },
          {
            name: "Test Series 2",
            series: {
              A: 10,
              B: 20,
              C: 30,
            },
          },
        ]}
        color="#ff1744"
        buttonText="Test Button Text"
        buttonLink="/"
      />,
    );

    expect(barChart.toJSON()).toMatchSnapshot();
    expect(barChart.root.findByType("h1").children[0]).toBe("TEST TITLE");
    expect(barChart.root.findAllByType(Chart).length).toBe(1);
    expect(barChart.root.findByType(Link).props.href).toBe("/");
    expect(barChart.root.findByType(Button).props.children).toBe(
      "Test Button Text",
    );
  });

  it("renders correctly with horizontal bar chart", () => {
    const barChart = TestRenderer.create(
      <BarChart
        title="Test Title"
        horizontal={true}
        dataList={[
          {
            name: "Test Series 1",
            series: {
              A: 10,
              B: 20,
              C: 30,
            },
          },
          {
            name: "Test Series 2",
            series: {
              A: 10,
              B: 20,
              C: 30,
            },
          },
        ]}
        color="#ff1744"
        buttonText="Test Button Text"
        buttonLink="/"
      />,
    );

    expect(barChart.toJSON()).toMatchSnapshot();
    expect(barChart.root.findByType("h1").children[0]).toBe("TEST TITLE");
    expect(barChart.root.findAllByType(Chart).length).toBe(1);
    expect(barChart.root.findByType(Link).props.href).toBe("/");
    expect(barChart.root.findByType(Button).props.children).toBe(
      "Test Button Text",
    );
  });

  it("renders correctly with no title", () => {
    const barChart = TestRenderer.create(
      <BarChart
        title={null}
        horizontal={false}
        dataList={[
          {
            name: "Test Series 1",
            series: {
              A: 10,
              B: 20,
              C: 30,
            },
          },
          {
            name: "Test Series 2",
            series: {
              A: 10,
              B: 20,
              C: 30,
            },
          },
        ]}
        color="#ff1744"
        buttonText="Test Button Text"
        buttonLink="/"
      />,
    );

    expect(barChart.toJSON()).toMatchSnapshot();
    expect(barChart.root.findAllByType("h1")).toHaveLength(0);
    expect(barChart.root.findAllByType(Chart).length).toBe(1);
    expect(barChart.root.findByType(Link).props.href).toBe("/");
    expect(barChart.root.findByType(Button).props.children).toBe(
      "Test Button Text",
    );
  });

  it("renders correctly with no button", () => {
    const barChart = TestRenderer.create(
      <BarChart
        title="Test Title"
        horizontal={false}
        dataList={[
          {
            name: "Test Series 1",
            series: {
              A: 10,
              B: 20,
              C: 30,
            },
          },
          {
            name: "Test Series 2",
            series: {
              A: 10,
              B: 20,
              C: 30,
            },
          },
        ]}
        color="#ff1744"
        buttonText={null}
        buttonLink="/"
      />,
    );

    expect(barChart.toJSON()).toMatchSnapshot();
    expect(barChart.root.findByType("h1").children[0]).toBe("TEST TITLE");
    expect(barChart.root.findAllByType(Chart).length).toBe(1);
    expect(barChart.root.findAllByType(Link)).toHaveLength(0);
    expect(barChart.root.findAllByType(Button)).toHaveLength(0);
  });
});

// LineChart tests.
describe("LineChart", () => {
  it("renders correctly with timeseries", () => {
    const lineChart = TestRenderer.create(
      <LineChart
        title="Test Title"
        dataList={[
          {
            name: "Test Series 1",
            series: [
              { x: "2019-01-01", y: 10 },
              { x: "2019-01-02", y: 20 },
              { x: "2019-01-03", y: 30 },
            ],
          },
          {
            name: "Test Series 2",
            series: [
              { x: "2019-01-01", y: 30 },
              { x: "2019-01-02", y: 10 },
              { x: "2019-01-03", y: 20 },
            ],
          },
        ]}
      />,
    );

    expect(lineChart.toJSON()).toMatchSnapshot();
    expect(lineChart.root.findByType("h1").children[0]).toBe("TEST TITLE");
    expect(lineChart.root.findAllByType(Chart).length).toBe(1);
  });
});

// PieChart tests.
describe("PieChart", () => {
  it("renders correctly with pie chart", () => {
    const pieChart = TestRenderer.create(
      <PieChart
        title="Test Title"
        donut={false}
        data={{
          A: 10,
          B: 20,
          C: 30,
        }}
      />,
    );

    expect(pieChart.toJSON()).toMatchSnapshot();
    expect(pieChart.root.findByType("h1").children[0]).toBe("TEST TITLE");
    expect(pieChart.root.findAllByType(Chart).length).toBe(1);
  });

  it("renders correctly with donut chart", () => {
    const pieChart = TestRenderer.create(
      <PieChart
        title="Test Title"
        donut={true}
        data={{
          A: 10,
          B: 20,
          C: 30,
        }}
      />,
    );

    expect(pieChart.toJSON()).toMatchSnapshot();
    expect(pieChart.root.findByType("h1").children[0]).toBe("TEST TITLE");
    expect(pieChart.root.findAllByType(Chart).length).toBe(1);
  });
});
