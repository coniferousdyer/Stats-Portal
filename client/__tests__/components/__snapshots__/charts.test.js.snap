// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChart renders correctly with horizontal bar chart 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 paper_container css-qh7kl5-MuiPaper-root"
>
  <h1
    className="title"
  >
    TEST TITLE
  </h1>
  <div />
  <div
    className="button_container"
  >
    <a
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root button css-sghohy-MuiButtonBase-root-MuiButton-root"
      href="/"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      Test Button Text
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </a>
  </div>
</div>
`;

exports[`BarChart renders correctly with no button 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 paper_container css-qh7kl5-MuiPaper-root"
>
  <h1
    className="title"
  >
    TEST TITLE
  </h1>
  <div />
</div>
`;

exports[`BarChart renders correctly with no title 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 paper_container css-qh7kl5-MuiPaper-root"
>
  <div />
  <div
    className="button_container"
  >
    <a
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root button css-sghohy-MuiButtonBase-root-MuiButton-root"
      href="/"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      Test Button Text
    </a>
  </div>
</div>
`;

exports[`BarChart renders correctly with vertical bar chart 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 paper_container css-qh7kl5-MuiPaper-root"
>
  <h1
    className="title"
  >
    TEST TITLE
  </h1>
  <div />
  <div
    className="button_container"
  >
    <a
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root button css-sghohy-MuiButtonBase-root-MuiButton-root"
      href="/"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      Test Button Text
    </a>
  </div>
</div>
`;

exports[`LineChart renders correctly with timeseries 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 paper_container css-qh7kl5-MuiPaper-root"
>
  <h1
    className="title"
  >
    TEST TITLE
  </h1>
  <div />
</div>
`;

exports[`PieChart renders correctly with donut chart 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 paper_container css-qh7kl5-MuiPaper-root"
>
  <h1
    className="title"
  >
    TEST TITLE
  </h1>
  <div />
</div>
`;

exports[`PieChart renders correctly with pie chart 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation5 paper_container css-qh7kl5-MuiPaper-root"
>
  <h1
    className="title"
  >
    TEST TITLE
  </h1>
  <div />
</div>
`;
